docs/fsharp/style-guide/conventions.md

```diff
@@ -0,0 +1,475 @@
+        state <- value = array.[i]
+        i <- i + 1
+    state
+```
+
+Calling this function multiple times does not change the underlying array, nor does it require you to maintain any mutable state by yourself. It is referentially transparent, even though almost every line of code inside of it uses mutation.
+
+### Consider encapsulating mutable data in classes
+
+The previous example used a single function to encapsulate operations using mutable data. This is not always sufficient for more complicated information. Consider the following sets of functions:
+
+```fsharp
+let addToClosureTable (t: Dictionary<_,_>) (a, b) =
+    if not (t.Contains(a)) then
+        t.[1] <- HashSet<_>(HashIdentity.Structural)
+    
+    t.[a].Add(b)
+
+let closureTableCount (t: Dictionary<_,_>) = t.Count
+
+let closureTableContains (t: Dictionary<_, HashSet<_>>) (a, b) =
+    t.ContainsKey(a) && t.[a].Contains(b)
+```
+
+This code is performant, but it exposes the mutation-based data structure which callers are responsible for maintaining. This can be wrapped inside of a class with no underlying members that can change:
+
+```fsharp
+/// The results of computing the LALR(1) closure of an LR(0) kernel
+type Closure1Table() =
+    let t = Dictionary<Item0, HashSet<TerminalIndex>>()
+
+    member __.Add(a, b) =
+        if not (t.Contains(a)) then
+            t.[1] <- HashSet<_>(HashIdentity.Structural)
+    
+        t.[a].Add(b)
+
+    member __.Count = t.Count
+
+    member __.Contains(a, b) = t.ContainsKey(a) && t.[a].Contains(b)
+```
+
+Because this class has no members which can change, and its binding is immutable, it is also effectively immutable. Additionally, it safely encapsulates the underlying mutation-based data structure. Classes are a powerful way to encapsulate data and routines which are mutation-based without exposing the details to callers.
+
+### Prefer `let mutable` to reference cells
+
+Reference cells can be a powerful mechanism for pointer arithemetic, but are generally not ideal for other performance-critical code. Consider the following function:
+
+```fsharp
+let kernels =
+    let acc = ref Set.empty
+
+    ProcessWorkList startKernels (fun kernel ->
+        if not ((!acc).Contains(kernel)) then
+            acc := (!acc).Add(kernel)
+        ...)
+
+    !acc |> Seq.toList
+```
+
+The use of a reference cell now "pollutes" all subsequent code with having to dereference and re-reference the underlying data. Instead, consider `let mutable`:
+
+```fsharp
+let kernels =
+    let mutable acc = Set.empty
+
+    ProcessWorkList startKernels (fun kernel ->
+        if not (acc.Contains(kernel) then
+            acc <- acc.Add(kernel)
+        ...)
+
+    acc |> Seq.toList
+```
+
+Aside from the single point of mutation in the middle of the lambda expression, all other code which works with `acc` no differently than if it were a normal `let`-bound immutable value. This will make it easier to change over time.
```

Comment:
which -> that

>all other code which works with  `acc` no differently than if it were a normal

all other code that touches `acc` can do so in a manner that is no different to the usage of a normal